src/dashboard/components/SignUp/SignUp.tsx

```diff
@@ -16,6 +16,7 @@ import { useMutation } from '@tanstack/react-query';
 import classes from './SignUp.module.css';
 import { signUp } from '@/auth';
 import { UserLogin } from '@/models/user.model';
+import { validatePassword, validateUsername } from '@/helper/validation';
 
 export function SignUp() {
   const form = useForm({
```

Comment:
This is a pretty weird way of doing it lmao, does mantine not offer a controlled `<Form />` component where you can control the states?

Also when I was referring to a stale state earlier I was moreso referring to the fact that the `mutation` object might become a stale reference (unless `useMutation` mutates in-place or if `validate` is not persisted between renders). Might be worth double checking this but this looks like its going to be a stale reference,

Reply:
Yh about forms, thats how mantine expects devs to use it (https://mantine.dev/form/use-form/)

The mutation object is fine, it mutates in-place so whenever a new mutation is triggered, the same reference still works. There was an issue initially where the sign-up form wouldn't submit after encountering an error, but I've just realised its because it was never passing the validate stage since I added mutation.isError (returns true if last mutation had an error) in there.

Reply:
Ah i see, honestly first time I've seen a component used like this before